### PR TITLE
Adding reservoir stratification scheme to MOSART-heat

### PR DESCRIPTION
**Pull request by wyigzaw**
The new module adds a reservoir stratification capability to MOSART-heat.
In doing so the effect of reservoirs on stream temperature seasonality and amplitude is better simulated.

- The stratification flag (rstraflag) should be set to "true" in "user_nl_mosart" for the scheme to be active

- Inflow/outflow are considered for the advection process

- A new reservoir dataset (geometry, number of layers, and outlet location) is also prepared.

- These reservoir parameters can be modified by users(for example, reservoir outflow location).
If required and/or additional reservoir data observation is available.

- User can also specify initial reservoir temperature(inittemp), vertical discretization points 
for reservoir geometry (ngeom), and	the maximum number of layers allowed (layers)

[BFB]

### CODE REVIEW

**Comment by rljacob** [comment]
Is this for v2 or v3?

**Comment by wyigzaw** [comment]
It is for v3.

File: cime/src/drivers/mct/shr/seq_flds_mod.F90 [hunk 1]
```diff
@@ -2014,6 +2014,15 @@ subroutine seq_flds_set(nmlfile, ID, infodata)
     units    = 'kg m-2 s-1'
     attname  = 'Flrl_rofsur'
     call metadata_set(attname, longname, stdname, units)
+	
+	! Cosine of Zenith angle (-)
```
Comment (bishtgautam) [inline cime/src/drivers/mct/shr/seq_flds_mod.F90:2018]: Add this new code within `if (trim(cime_model) == 'e3sm') then`

File: cime/src/drivers/mct/main/prep_rof_mod.F90 [hunk 2]
```diff
@@ -447,6 +447,9 @@ subroutine prep_rof_merge(l2x_r, a2x_r, fractions_r, x2r_r, cime_model)
     integer, save :: index_x2r_Faxa_swvdr
     integer, save :: index_x2r_Faxa_swvdf
     integer, save :: index_x2r_Faxa_lwdn
+	
+	integer, save :: index_l2x_coszen_str
```
Comment (bishtgautam) [inline cime/src/drivers/mct/main/prep_rof_mod.F90:451]: Please indent the line

File: components/clm/src/cpl/clm_cpl_indices.F90 [hunk 3]
```diff
@@ -175,7 +175,7 @@ subroutine clm_cpl_indices_set( )
     index_l2x_Flrl_demand   = mct_avect_indexra(l2x,'Flrl_demand')
     index_l2x_Flrl_Tqsur    = mct_avect_indexra(l2x,'Flrl_Tqsur')
     index_l2x_Flrl_Tqsub    = mct_avect_indexra(l2x,'Flrl_Tqsub')
-
+	index_l2x_coszen_str	= mct_avect_indexra(l2x,'coszen_str')
```
Comment (bishtgautam) [inline components/clm/src/cpl/clm_cpl_indices.F90:178]: Please indent the line

File: components/clm/src/cpl/lnd_import_export.F90 [hunk 4]
```diff
@@ -1405,7 +1405,7 @@ subroutine lnd_export( bounds, lnd2atm_vars, lnd2glc_vars, l2x)
        endif
        l2x(index_l2x_Flrl_Tqsur,i)  = lnd2atm_vars%Tqsur_grc(g)
        l2x(index_l2x_Flrl_Tqsub,i)  = lnd2atm_vars%Tqsub_grc(g)
-
+	   l2x(index_l2x_coszen_str,i) = lnd2atm_vars%coszen_str(g)
```
Comment (bishtgautam) [inline components/clm/src/cpl/lnd_import_export.F90:1408]: Fix indentation

File: components/mosart/src/riverroute/RtmHistFlds.F90 [hunk 5]
```diff
@@ -213,6 +213,21 @@ subroutine RtmHistFldsInit()
            avgflag='A', long_name='Water temperature of main channels', &
            ptr_rof=rtmCTL%templand_Tchanr_nt1)         
     end if     
+
+    if (wrmflag .and. heatflag .and. rstraflag) then
+		call RtmHistAddfld (fname='RSRV_SURF', units='Kelvin',  &
+			avgflag='A', long_name='Reservoir surface temperature', &
+			ptr_rof=WRMUnit%resrv_surf)	
+        ! call RtmHistAddfld (fname='RSRV_OUT', units='Kelvin',  &
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmHistFlds.F90:221]: Remove commented out lines

File: components/mosart/src/riverroute/RtmMod.F90 [hunk 6]
```diff
@@ -258,6 +258,10 @@ subroutine Rtmini(rtm_active,flood_active)
     do_rtmflood = .false.
     ice_runoff  = .true.
     wrmflag     = .false.
+	rstraflag   = .false.
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmMod.F90:261]: Fix indentation

File: components/mosart/src/riverroute/RtmMod.F90 [hunk 7]
```diff
@@ -340,6 +344,10 @@ subroutine Rtmini(rtm_active,flood_active)
     call mpi_bcast (wrmflag,        1, MPI_LOGICAL, 0, mpicom_rof, ier)
     call mpi_bcast (sediflag,       1, MPI_LOGICAL, 0, mpicom_rof, ier)
     call mpi_bcast (heatflag,       1, MPI_LOGICAL, 0, mpicom_rof, ier)
+	call mpi_bcast (rstraflag,      1, MPI_LOGICAL, 0, mpicom_rof, ier)
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmMod.F90:347]: Fix indentation

File: components/mosart/src/riverroute/RtmMod.F90 [hunk 8]
```diff
@@ -1542,6 +1550,9 @@ subroutine Rtmini(rtm_active,flood_active)
        endif
        call t_startf('mosarti_wrm_init')
     end if
+	if (wrmflag .and. heatflag .and. rstraflag) then
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmMod.F90:1553]: Fix indentation

File: components/mosart/src/riverroute/RtmMod.F90 [hunk 9]
```diff
@@ -3895,6 +3906,12 @@ subroutine MOSART_init
              allocate(TPara%t_mu(begr:endr))
            TPara%t_mu = 0.5278_r8
         end if
+		
+		! if(rstraflag) then
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmMod.F90:3910]: Fix indentation and remove commented out lines

File: components/mosart/src/riverroute/RtmMod.F90 [hunk 10]
```diff
@@ -4049,7 +4066,237 @@ subroutine MOSART_init
...
+	
+		do iunit = rtmCTL%begr,rtmCTL%endr	
+		
+			damID = WRMUnit%INVicell(iunit)
+			if (damID>0.0_r8 .and. WRMUnit%d_ns(damID) >= 1) then ! .and. WRMUnit%Depth(damID) > 0.0_r8 .and. WRMUnit%Height(damID) > 0.0_r8
+				if (WRMUnit%Depth(damID) <= 2.0_r8)WRMUnit%Depth(damID) = 2._r8
+				if (WRMUnit%Height(damID) <= 2.0_r8)WRMUnit%Height(damID) = WRMUnit%Depth(damID)
+				WRMUnit%d_resrv(damID) = 0.95_r8*WRMUnit%Height(damID)
+				WRMUnit%h_resrv(damID) = 0.95_r8*WRMUnit%Height(damID)
+				! if(WRMUnit%d_resrv(damID) >= 100._r8 .and. WRMUnit%d_ns(damID)>=10)WRMUnit%d_ns(damID)= WRMUnit%d_ns(damID) + max((nlayers - 30),0) ! affects only large/deep reservoirs
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmMod.F90:4105]: Remove commented out lines

File: components/mosart/src/riverroute/RtmMod.F90 [hunk 11]
```diff
@@ -4049,7 +4066,237 @@ subroutine MOSART_init
...
+				
+				if (WRMUnit%A_str(damID)<= 0._r8) WRMUnit%A_str(damID)= 1._r8
+				
+				if (WRMUnit%A_str(damID)<= 2._r8) then
+					C_a  = 2.0_r8*WRMUnit%C_as(damID)
+					! gm_j  = 1
+					! WRMUnit%d_ns(damID) = 1
+				end if
+				
+				! if (a_sur/s_tin < 0.01_r8 .and. (WRMUnit%V_errs(damID) > 25._r8 .or. WRMUnit%A_errs(damID) > 25._r8)) gm_j = 1.0_r8
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmMod.F90:4152]: remove commented out lines

File: components/mosart/src/riverroute/MOSART_physics_mod.F90 [hunk 12]
```diff
@@ -388,7 +389,10 @@ subroutine Euler
                 ! moved out of loop
                    if ( ctlSubwWRM%RegulationFlag>0 ) then
                       call Regulation(iunit, localDeltaT)
-                      if (heatflag) then
+                      if (heatflag .and. rstraflag) then
+						  call stratification(iunit, localDeltaT,nt)
```
Comment (hydrotian) [inline components/mosart/src/riverroute/MOSART_physics_mod.F90:393]: Change tab to space. Please check other places too.

**Comment by bishtgautam** [comment]
This has been replaced by #3735

**Pull request closed without merging**